### PR TITLE
Delay user errors in Dir_contents

### PR DESCRIPTION
**Pull request by rgrinberg**
User errors should be presented only when the rule is used to build a
target. This commit moves towards that by allowing callers to setup
rules prefixed with the exceptions.

### CODE REVIEW

**Comment by snowleopard** [comment]
Can you add an example/test that demonstrates the problem you are solving?

I had a brief look at the code. The `ok_exn` popping all over the place are a bit annoying... is this unavoidable?

File: src/dune/ml_sources.ml [hunk 1]
```diff
@@ -237,26 +247,19 @@ let make_lib_modules (d : _ Dir_with_dune.t) ~lookup_vlib ~(lib : Library.t)
         (* can't happen because this library is defined using the current stanza *)
         |> Option.value_exn
       in
-      (* diml: this [Result.ok_exn] means that if the user writes an invalid
-         [implements] field, we will get an error immediately even if the
-         library is not built. We should change this to carry the [Or_exn.t] a
-         bit longer. *)
-      let vlib =
-        Result.ok_exn
```
Comment (rgrinberg) [inline src/dune/ml_sources.ml]: @snowleopard this is the error that we are trying to delay. The issue is that the absence of a virtual library means that loading the rules in an entire directory will be broken.

The PR doesn't yet do that, but it goes in the right direction. The callers of `Dir_contents` should be turning `Error` values into `Build.fail`.

Comment (snowleopard) [reply src/dune/ml_sources.ml]: It would be nice to add a test exposing this undesirable behaviour, so that this PR fixes it.

Comment (snowleopard) [reply src/dune/ml_sources.ml:245]: (Otherwise, we will likely regress back at some point.)

**Pull request closed without merging**